src/plugins/directoryvalue/directoryvalue_delegate.cpp

```diff
@@ -0,0 +1,524 @@
+/**
+ * @file
+ *
+ * @brief Delegate implementation for the `directoryvalue` plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include <numeric>
+
+#include <kdbassert.h>
+#include <kdbease.h>
+#include <kdblogger.h>
+
+#include "directoryvalue_delegate.hpp"
+
+using std::accumulate;
+using std::ignore;
+using std::make_pair;
+using std::pair;
+using std::range_error;
+using std::string;
+using std::tie;
+
+// -- Functions ----------------------------------------------------------------------------------------------------------------------------
+
+namespace elektra
+{
+using CppKey = kdb::Key;
+
+/**
+ * @brief This function splits the given keyset into directory leaves (marked with `DIRECTORY_POSTFIX`) and other keys.
+ *
+ * @param input The function searches for directory leaves in this key set.
+ *
+ * @return A pair of key sets, where the first key set contains all directory leaves and the second key set contains all other keys
+ */
+pair<CppKeySet, CppKeySet> splitDirectoryLeavesOther (CppKeySet const & input)
+{
+	CppKeySet directoryLeaves;
+	CppKeySet other;
+
+	for (auto key : input)
+	{
+		if (key.getBaseName () == DIRECTORY_POSTFIX)
+		{
+			directoryLeaves.append (key);
+		}
+		else
+		{
+			other.append (key);
+		}
+	}
+	return make_pair (directoryLeaves, other);
+}
+
+/**
+ * @brief This function splits the given keyset into array leaves (first element of an array parent) and other keys.
+ *
+ * @param input The function searches for array leaves in this key set.
+ *
+ * @return A pair of key sets, where the first key set contains all array leaves and the second key set contains all other keys
+ */
+pair<CppKeySet, CppKeySet> splitArrayLeavesOther (CppKeySet const & arrayParents, CppKeySet const & keys)
+{
+	bool isFirstElement = false;
+	CppKeySet firstElements;
+	CppKeySet others;
+
+	for (auto key : keys)
+	{
+		bool isArrayLeaf = isFirstElement && key.isString () && key.getStringSize () > arrayValuePrefixSize &&
+				   strncmp (key.getString ().c_str (), ARRAY_VALUE_PREFIX, arrayValuePrefixSize) == 0;
+		(isArrayLeaf ? firstElements : others).append (key);
+		isFirstElement = arrayParents.lookup (key);
+	}
+
+	return make_pair (firstElements, others);
+}
+
+/**
+ * @brief This function removes the basename (last level) from the given keys.
+ *
+ * @param keys This parameter contains the keys for which this function removes the basename.
+ *
+ * @return A copy of the input, where the last level of each key was removed
+ */
+CppKeySet removeBaseName (CppKeySet const & keys)
+{
+	CppKeySet directories;
+
+	for (auto key : keys)
+	{
+		CppKey directory = key.dup ();
+		keySetBaseName (*directory, 0);
```

Comment:
Why using C here? setBaseName exists?

Reply:
I tried to use `directory.setBaseName (0)`, but this function call causes segmentation faults. As far as I can tell there is no option to remove the basename in the C++ API. At least I only found the C++ method:

```cpp
void setBaseName (const std::string & baseName);
```

, which takes a string as argument. If I add the following function:

```cpp
inline void Key::setBaseName (char const * baseName)
{
	if (ckdb::keySetBaseName (getKey (), baseName) == -1)
	{
		throw KeyInvalidName ();
	}
}
```

everything works fine.

Reply:
Thank you, there were various errors in this documentation. I tried to fix them all, please open an issue if there are problems in the docu.

I created a new issue #2182 to add a delBaseName(). Please also open issues if the C++ binding does not allow you to do something which is possible in the C API.